examples/swap-tracks/main.go

```diff
@@ -0,0 +1,196 @@
+	peerConnection, err := api.NewPeerConnection(config)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create Track that we send video back to browser on
+	outputTrack, err := peerConnection.NewTrack(videoCodecs[0].PayloadType, rand.Uint32(), "video", "pion")
+	if err != nil {
+		panic(err)
+	}
+
+	// Add this newly created track to the PeerConnection
+	if _, err = peerConnection.AddTrack(outputTrack); err != nil {
+		panic(err)
+	}
+
+	// In addition to the implicit transceiver added by the track, we add two more
+	// for the other tracks
+	_, err = peerConnection.AddTransceiverFromKind(webrtc.RTPCodecTypeVideo,
+		webrtc.RtpTransceiverInit{Direction: webrtc.RTPTransceiverDirectionRecvonly})
+	if err != nil {
+		panic(err)
+	}
+	_, err = peerConnection.AddTransceiverFromKind(webrtc.RTPCodecTypeVideo,
+		webrtc.RtpTransceiverInit{Direction: webrtc.RTPTransceiverDirectionRecvonly})
+	if err != nil {
+		panic(err)
+	}
+
+	// Set the remote SessionDescription
+	err = peerConnection.SetRemoteDescription(offer)
+	if err != nil {
+		panic(err)
+	}
+
+	// Which track is currently being handled
+	currTrack := 0
+	// The total number of tracks
+	trackCount := 0
+	// The channel of packets with a bit of buffer
+	packets := make(chan *rtp.Packet, 60)
+
+	// Set a handler for when a new remote track starts
+	peerConnection.OnTrack(func(track *webrtc.Track, receiver *webrtc.RTPReceiver) {
+		fmt.Printf("Track has started, of type %d: %s \n", track.PayloadType(), track.Codec().Name)
+		trackNum := trackCount
+		trackCount++
+		// The last timestamp so that we can change the packet to only be the delta
+		var lastTimestamp uint32
+		// Whether this track is the one currently sending to the channel (on change
+		// of this we send a PLI to have the entire picture updated)
+		var onMe bool
+		for {
+			// Read RTP packets being sent to Pion
+			rtp, readErr := track.ReadRTP()
+			if readErr != nil {
+				panic(readErr)
+			}
+
+			// Change the timestamp to only be the delta
+			oldTimestamp := rtp.Timestamp
+			if lastTimestamp == 0 {
+				rtp.Timestamp = 0
+			} else {
+				rtp.Timestamp -= lastTimestamp
+			}
+			lastTimestamp = oldTimestamp
+
+			// Check if this is the current track
+			if currTrack == trackNum {
+				// If just switched to this track, send PLI to get picture refresh
+				if !onMe {
+					onMe = true
+					writeErr := peerConnection.WriteRTCP([]rtcp.Packet{&rtcp.PictureLossIndication{MediaSSRC: track.SSRC()}})
+					if writeErr != nil {
+						fmt.Println(writeErr)
+					}
+				}
+				packets <- rtp
+			} else {
+				onMe = false
+			}
+		}
+	})
+	// Set the handler for ICE connection state and update chan if connected
+	connected := make(chan struct{}, 1)
+	peerConnection.OnICEConnectionStateChange(func(connectionState webrtc.ICEConnectionState) {
+		fmt.Printf("Connection State has changed %s \n", connectionState.String())
+		// Non-blocking just in case already sent
+		if connectionState == webrtc.ICEConnectionStateConnected {
```

Comment:
I would do a context, I think you can call done as many times as you want?

Reply:
There are a ton of ways to skin this cat. Could use a context cancel func, could use an atomic value, could use `sync.Once` with simple chan with `Do(func() { close(doneCh) })`, could use `sync.Once` with `sync.WaitGroup` and `Do(wg.Done)`, could use `sync.Cond`, etc, etc.

If it's much clearer with context cancellation, I can use that (I tend not to use it as a notifier as it does read as clear as context's common uses).